### PR TITLE
fix: prevent adding/updating a provider with an empty or existing id

### LINKED ISSUE #598: Data Provider without ID
You can create a data provider without an ID and it works, but when you want to change it (put an id into) or want to delete you can't.
Output:
```
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>Cannot DELETE /providers/</pre>
</body>
</html>
```

### PR DESCRIPTION
**Pull request by sbender9**
Closes #598

### CODE REVIEW

**Comment by tkurki** [comment]
I would have written a failing api test first.

File: lib/config/config.js [hunk 1]
```diff
@@ -246,8 +249,12 @@ function readSettingsFile (app) {
 }
 
 function writeSettingsFile (app, settings, cb) {
-  const settingsPath = getSettingsFilename(app)
-  fs.writeFile(settingsPath, JSON.stringify(settings, null, 2), cb)
+  if (!disableWriteSettings) {
+    const settingsPath = getSettingsFilename(app)
+    fs.writeFile(settingsPath, JSON.stringify(settings, null, 2), cb)
+  } else {
+    cb()
```
Comment (tkurki) [inline lib/config/config.js:256]: This should be probably be cb(”saving settings disabled”)?

**Comment by tkurki** [comment]
Put and post for providers have quite a bit of duplication and for example only one has the security check. Time to refactor: extract the shared logic.

File: lib/config/config.js (hunk 1 above)
Comment (sbender9) [reply lib/config/config.js:256]: But then the tests will fail, the PUT/POSTs fail because they think there was an error saving

**Pull request merged**